cpp/src/arrow/array/concatenate_test.cc

```diff
@@ -364,27 +364,89 @@ TEST_F(ConcatenateTest, DictionaryTypeNullSlots) {
+  ASSERT_OK_AND_ASSIGN(auto array,
+                       DenseUnionArray::Make(*type_ids, *offsets, {child_array}));
+  ArrayVector arrays({array, array});
+  ASSERT_RAISES(Invalid, Concatenate(arrays, default_memory_pool()));
+
+  // Length overflow
+  auto type_ids_ok = ArrayFromJSON(int8(), "[0]");
+  auto offsets_ok = ArrayFromJSON(int32(), "[0]");
+  auto child_array_overflow =
+      this->rng_.ArrayOf(null(), std::numeric_limits<int32_t>::max() - 1, 0.0);
+  ASSERT_OK_AND_ASSIGN(
+      auto array_overflow,
+      DenseUnionArray::Make(*type_ids_ok, *offsets_ok, {child_array_overflow}));
+  ArrayVector arrays_overflow({array_overflow, array_overflow});
+  ASSERT_RAISES(Invalid, Concatenate(arrays_overflow, default_memory_pool()));
+}
+
+TEST_F(ConcatenateTest, DenseUnionType) {
+  auto type_ids = ArrayFromJSON(int8(), "[0, 0, 1, 1, 0, 1]");
+  auto offsets = ArrayFromJSON(int32(), "[0, 1, 0, 1, 2, 2]");
+  auto child_one = ArrayFromJSON(boolean(), "[true, true, false]");
+  auto child_two = ArrayFromJSON(int8(), "[1, 2, 3]");
+  ASSERT_OK_AND_ASSIGN(
+      auto array, DenseUnionArray::Make(*type_ids, *offsets, {child_one, child_two}));
+  ASSERT_OK(array->ValidateFull());
+
+  ArrayVector arrays({array, array});
+  ASSERT_OK_AND_ASSIGN(auto concat_arrays, Concatenate(arrays, default_memory_pool()));
+  ASSERT_OK(concat_arrays->ValidateFull());
+
+  auto type_ids_expected = ArrayFromJSON(int8(), "[0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1]");
+  auto offsets_expected = ArrayFromJSON(int32(), "[0, 1, 0, 1, 2, 2, 3, 4, 3, 4, 5, 5]");
+  auto child_one_expected =
+      ArrayFromJSON(boolean(), "[true, true, false, true, true, false]");
+  auto child_two_expected = ArrayFromJSON(int8(), "[1, 2, 3, 1, 2, 3]");
+  ASSERT_OK_AND_ASSIGN(auto expected,
+                       DenseUnionArray::Make(*type_ids_expected, *offsets_expected,
+                                             {child_one_expected, child_two_expected}));
+  ASSERT_OK(expected->ValidateFull());
+  AssertArraysEqual(*expected, *concat_arrays);
+
+  ArrayVector sliced_arrays({array->Slice(1, 4), array});
+  ASSERT_OK_AND_ASSIGN(auto concat_sliced_arrays,
+                       Concatenate(sliced_arrays, default_memory_pool()));
+  ASSERT_OK(concat_sliced_arrays->ValidateFull());
+
+  auto type_ids_sliced = ArrayFromJSON(int8(), "[0, 1, 1, 0, 0, 0, 1, 1, 0, 1]");
+  auto offsets_sliced = ArrayFromJSON(int32(), "[1, 0, 1, 2, 3, 4, 3, 4, 5, 5]");
+  auto child_one_sliced =
+      ArrayFromJSON(boolean(), "[true, true, false, true, true, false]");
+  auto child_two_sliced = ArrayFromJSON(int8(), "[1, 2, 3, 1, 2, 3]");
+  ASSERT_OK_AND_ASSIGN(auto expected_sliced,
+                       DenseUnionArray::Make(*type_ids_sliced, *offsets_sliced,
+                                             {child_one_sliced, child_two_sliced}));
+  ASSERT_OK(expected_sliced->ValidateFull());
+  AssertArraysEqual(*expected_sliced, *concat_sliced_arrays);
```

Comment:
I think for for most of these arrays it's acceptable and more readable to come straight from JSON
```suggestion
  AssertArraysEqual(*ArrayFromJSON(dense_union({field("", boolean()), field("", int8())}), R"([
    [0, true],
    [1, 1],
    [1, 2],
    [0, true],
    [0, false],
    [0, true],
    [1, 3],
    [1, 1],
    [0, true],
    [1, 3]
  ])"), *concat_sliced_arrays);
```

The only case where you *need* to construct the dense union array with explicit children is when you're making an array with sliced children